google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1alpha2/WriterCache.java

```diff
@@ -0,0 +1,242 @@
+  @VisibleForTesting
+  public static WriterCache getTestInstance(
+      BigQueryWriteClient stub, int maxTableEntry, SchemaCompact compact) {
+    return new WriterCache(stub, maxTableEntry, compact);
+  }
+
+  /** Returns an entry with {@code StreamWriter} and expiration time in millis. */
+  private String CreateNewStream(String tableName) {
+    Stream.WriteStream stream =
+        Stream.WriteStream.newBuilder().setType(Stream.WriteStream.Type.COMMITTED).build();
+    stream =
+        stub.createWriteStream(
+            Storage.CreateWriteStreamRequest.newBuilder()
+                .setParent(tableName)
+                .setWriteStream(stream)
+                .build());
+    LOG.info("Created write stream:" + stream.getName());
+    return stream.getName();
+  }
+
+  StreamWriter CreateNewWriter(String streamName)
+      throws IllegalArgumentException, IOException, InterruptedException {
+    return StreamWriter.newBuilder(streamName)
+        .setChannelProvider(stub.getSettings().getTransportChannelProvider())
+        .setCredentialsProvider(stub.getSettings().getCredentialsProvider())
+        .setExecutorProvider(stub.getSettings().getExecutorProvider())
+        .build();
+  }
+  /**
+   * Gets a writer for a given table with a given user schema from global cache.
+   *
+   * @param tableName
+   * @param userSchema
+   * @return
+   * @throws Exception
+   */
+  public StreamWriter getTableWriter(String tableName, Descriptor userSchema)
+      throws IllegalArgumentException, IOException, InterruptedException {
+    Matcher matcher = tablePattern.matcher(tableName);
+    if (!matcher.matches()) {
+      throw new IllegalArgumentException("Invalid table name: " + tableName);
+    }
+
+    String streamName = null;
+    Boolean streamExpired = false;
+    StreamWriter writer = null;
+    LRUCache<Descriptor, StreamWriter> tableEntry = null;
+
+    synchronized (this) {
+      tableEntry = writerCache.get(tableName);
+      if (tableEntry != null) {
+        writer = tableEntry.get(userSchema);
+        if (writer != null) {
+          if (!writer.expired()) {
+            return writer;
+          } else {
+            writer.close();
+          }
+        }
+        compact.check(tableName, userSchema);
+        streamName = CreateNewStream(tableName);
+        writer = CreateNewWriter(streamName);
+        tableEntry.put(userSchema, writer);
+      } else {
+        compact.check(tableName, userSchema);
+        streamName = CreateNewStream(tableName);
+        tableEntry = new LRUCache<Descriptor, StreamWriter>(MAX_WRITERS_PER_TABLE);
+        writer = CreateNewWriter(streamName);
+        tableEntry.put(userSchema, writer);
+        writerCache.put(tableName, tableEntry);
+      }
+    }
+
+    return writer;
+  }
+
+  @VisibleForTesting
+  public int cachedTableCount() {
+    synchronized (writerCache) {
+      return writerCache.getCurrentSize();
+    }
+  }
+
+  private static class LRUCache<K, V> {
```

Comment:
Have you considered using [Guava](https://github.com/google/guava/wiki/CachesExplained)? We should not be writing our own LRU cahce.